### PR TITLE
Separate Wasmstore from normal db: initial

### PR DESCRIPTION
**Pull request by tsahee**

### CODE REVIEW

File: arbitrum/recordingdb.go [hunk 1]
```diff
@@ -57,10 +58,10 @@ func (db *RecordingKV) Get(key []byte) ([]byte, error) {
 		res, err = db.diskDb.Get(key)
 	} else if ok, _ := rawdb.IsActivatedAsmKey(key); ok {
 		// Arbitrum: the asm is non-consensus
-		return db.diskDb.Get(key)
+		return db.wasmDb.Get(key)
 	} else if ok, _ := rawdb.IsActivatedModuleKey(key); ok {
 		// Arbitrum: the module is non-consensus (only its hash is)
-		return db.diskDb.Get(key)
+		return db.wasmDb.Get(key)
```
Comment (PlasmaPower) [inline arbitrum/recordingdb.go:64]: Will these ever be called? Shouldn't they be called on the wasm database instead?

Comment (tsahee) [reply arbitrum/recordingdb.go]: moved some things around and now it is indeed not needed

File: core/state/journal_arbitrum.go [hunk 2]
```diff
@@ -41,8 +41,11 @@ type EvictWasm struct {
 }
 
 func (ch EvictWasm) revert(s *StateDB) {
-	asm := s.GetActivatedAsm(ch.ModuleHash) // only happens in native mode
-	CacheWasmRust(asm, ch.ModuleHash, ch.Version, ch.Debug)
+	asm, err := s.TryGetActivatedAsm(ch.ModuleHash) // only happens in native mode
+	if err == nil && len(asm) == 0 {
```
Comment (PlasmaPower) [inline core/state/journal_arbitrum.go:45]: I don't think this check is right. Maybe it should be `if err != nil || len(asm) == 0 {`?

File: core/state/journal_arbitrum.go [hunk 3]
```diff
@@ -41,8 +41,11 @@ type EvictWasm struct {
...
+		//if we failed to get it - it's not in the current rust cache
+		CacheWasmRust(asm, ch.ModuleHash, ch.Version, ch.Debug)
```
Comment (PlasmaPower) [inline core/state/journal_arbitrum.go:47]: I'm not really sure what's happening here. If we failed to get asm, why are we caching that same asm? Don't we need to regenerate it or something?

File: core/state/statedb_arbitrum.go [hunk 4]
```diff
@@ -237,9 +234,13 @@ func (s *StateDB) StartRecording() {
 }
 
 func (s *StateDB) RecordProgram(moduleHash common.Hash) {
+	asm, err := s.TryGetActivatedAsm(moduleHash)
+	if err != nil {
+		log.Crit("can't find activated wasm while recording", "modulehash", moduleHash)
```
Comment (PlasmaPower) [inline core/state/statedb_arbitrum.go:239]: Is the program guaranteed to be in cache here? I don't think we can trust that it's in the database, e.g. if the tx reverts.

File: internal/ethapi/api.go [hunk 5]
```diff
@@ -962,12 +962,17 @@ type OverrideAccount struct {
 // StateOverride is the collection of overridden accounts.
 type StateOverride map[common.Address]OverrideAccount
 
+var arbStorageAddress = common.HexToAddress("0xA4B05FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF")
```
Comment (PlasmaPower) [inline internal/ethapi/api.go:965]: There's already `types.ArbosStateAddress`

**Review by PlasmaPower — CHANGES_REQUESTED** [review]

File: core/state/journal_arbitrum.go (hunk 3 above)
Comment (tsahee) [reply core/state/journal_arbitrum.go:47]: yeah, it's wrong. We shoul only cache it if we succeed.

File: core/state/statedb_arbitrum.go (hunk 4 above)
Comment (tsahee) [reply core/state/statedb_arbitrum.go:239]: yes. It's called from arbos/programs/native.go:callProgram. Before we get to it we take the asm and module and store them at least to the statedb.

**Review by PlasmaPower — APPROVED** [review]
LGTM

**Pull request merged**